pkg/services/ngalert/api/api_ruler_validation.go

```diff
@@ -145,6 +100,85 @@ func validateRuleNode(
 	return &newAlertRule, nil
 }
 
+func validateAlertingRule(in *apimodels.PostableExtendedRuleNode, newRule *ngmodels.AlertRule, canPatch bool) error {
```

Comment:
The side effect modification of `newRule` isn't super obvious without reading the whole function, WDYT about this and `validateRecordingRule` either returning the validated rule or adding docstrings to mention that field is modified?

Reply:
Yeah that's fair. The main reason I didn't return a new object was that it wouldn't be fully formed yet. The result of the function wouldn't be a usable alert rule, it would be missing many shared fields like Title. I thought this captured the idea better that "this function has to be used alongside other things"

I thought about instead inverting this - where the shared validations are also extracted to a function. Then, the top level would call just `validateAlertingRule`, which would then call `sharedValidations` and then all the alert specific validations. Then, it could more stably return a fully-formed rule.

Reply:
I did the comment approach for now, and I also renamed the functions to `validate<X>RuleFields` to better express that they only validate a subset of the rule.

I think that's probably enough, if you disagree i could go the fully separated route with a function for common validation as well. The cost of this is more indirection in the code, though.

Reply:
Discussed offline, pushed another update which modifies without a reference and returns the result. This also changed the call site a bit